Dotty fails to typecheck scala/compat/java8

```
[error] ./src/./scala/compat/java8/JFunction0.java:9: error: not found: type R
[error] public interface JFunction0<R> extends scala.Function0<R> {
[error]                                                        ^
```

--- a/src/dotty/tools/dotc/parsing/JavaParsers.scala
+++ b/src/dotty/tools/dotc/parsing/JavaParsers.scala
@@ -114,7 +114,7 @@ object JavaParsers {
 
     def makeTemplate(parents: List[Tree], stats: List[Tree], tparams: List[TypeDef], needsDummyConstr: Boolean) = {
       def pullOutFirstConstr(stats: List[Tree]): (Tree, List[Tree]) = stats match {
-        case (meth: DefDef) :: rest if meth.name.isConstructorName => (meth, rest)
+        case (meth: DefDef) :: rest if meth.name == CONSTRUCTOR => (meth, rest)
         case first :: rest =>
           val (constr, tail) = pullOutFirstConstr(rest)
           (constr, first :: tail)

--- a/src/dotty/tools/dotc/typer/FrontEnd.scala
+++ b/src/dotty/tools/dotc/typer/FrontEnd.scala
@@ -28,7 +28,8 @@ class FrontEnd extends Phase {
     unit.untpdTree =
       if (unit.isJava) new JavaParser(unit.source).parse()
       else new Parser(unit.source).parse()
-    typr.println("parsed:\n" + unit.untpdTree.show)
+    val printer = if (ctx.settings.Xprint.value.contains("parser")) default else typr
+    printer.println("parsed:\n" + unit.untpdTree.show)
   }
 
   def enterSyms(implicit ctx: Context) = monitor("indexing") {

--- a/tests/pos/i851.java
+++ b/tests/pos/i851.java
@@ -1,0 +1,8 @@
+interface I<T> {}
+
+interface J<T> extends I<T> {
+    default void $init$() {
+    };
+}
+
+class C<T> extends J<T> {}

